src/hotspot/share/runtime/handshake.hpp

```diff
@@ -39,74 +40,75 @@ class JavaThread;
 // the target thread in a blocked state. A handshake can be performed with a
 // single JavaThread as well. In that case, the callback is executed either
 // by the target JavaThread itself or, depending on whether the operation is
-// a direct handshake or not, by the JavaThread that requested the handshake
-// or the VMThread respectively.
-class HandshakeClosure : public ThreadClosure {
+// a single target/direct handshake or not, by the JavaThread that requested the
+// handshake or the VMThread respectively.
+class HandshakeClosure : public ThreadClosure, public CHeapObj<mtThread> {
   const char* const _name;
  public:
   HandshakeClosure(const char* name) : _name(name) {}
-  const char* name() const {
-    return _name;
-  }
+  virtual ~HandshakeClosure() {}
+  const char* name() const    { return _name; }
+  virtual bool is_asynch()    { return false; };
   virtual void do_thread(Thread* thread) = 0;
 };
 
+class AsynchHandshakeClosure : public HandshakeClosure {
+ public:
+   AsynchHandshakeClosure(const char* name) : HandshakeClosure(name) {}
+   virtual ~AsynchHandshakeClosure() {}
+   virtual bool is_asynch()          { return true; }
+};
+
 class Handshake : public AllStatic {
  public:
   // Execution of handshake operation
-  static void execute(HandshakeClosure* hs_cl);
-  static bool execute(HandshakeClosure* hs_cl, JavaThread* target);
-  static bool execute_direct(HandshakeClosure* hs_cl, JavaThread* target);
+  static void execute(HandshakeClosure*       hs_cl);
+  static void execute(HandshakeClosure*       hs_cl, JavaThread* target);
+  static void execute(AsynchHandshakeClosure* hs_cl, JavaThread* target);
 };
 
 // The HandshakeState keeps track of an ongoing handshake for this JavaThread.
-// VMThread/Handshaker and JavaThread are serialized with semaphore _processing_sem
-// making sure the operation is only done by either VMThread/Handshaker on behalf
-// of the JavaThread or by the target JavaThread itself.
+// VMThread/Handshaker and JavaThread are serialized with _lock making sure the
+// operation is only done by either VMThread/Handshaker on behalf of the
+// JavaThread or by the target JavaThread itself.
 class HandshakeState {
   JavaThread* _handshakee;
-  HandshakeOperation* volatile _operation;
-  HandshakeOperation* volatile _operation_direct;
-
-  Semaphore _handshake_turn_sem;  // Used to serialize direct handshakes for this JavaThread.
-  Semaphore _processing_sem;
-  bool _thread_in_process_handshake;
+  FilterQueue<HandshakeOperation*> _queue;
+  Mutex _lock;
+  Thread* _active_handshaker;
```

Comment:
The order of members matter since C++ initialize them in declared order.
My opinion when changing this was that it was easier to read when passing the only argument to the first member being initialized, thus _handshakee must be first member.

But I should init _active_handshaker in constructor, so added that and lined-up.

So before I do any such change please reflect over how the constructor will look like.

Reply:
I don't understand, you'd have to rearrange the initializers in the constructor too, but I don't see any order dependance.  Moving over _lock  helps, so this is fine.

Reply:
You want a cosmetic change in the member declaration.
I'm saying the constructor will look worse.

Im asking if you want to trade a worse constructor for that?
(all here is extremely subjective :) )

Reply:
It is subjective.  What you have is fine.